File: src/ChessGame/ChessGame.Domain/Entitites/Board.cs
using ChessGame.Domain.Entitites.Interfaces;
using ChessGame.Domain.Entitites.Pieces;
using ChessGame.Domain.Entitites.Pieces.Base;
using ChessGame.Domain.ValueObjects;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ChessGame.Domain.Entitites
{
    public class Board
    {
        public Board(Game game)
        {
            Game = game;
            Size = new Size(8, 8);
        }

        public Size Size { get; }

        public Game Game { get; }

        public List<IPiece> GetPieces() => Game.WhitesPlayer.Pieces.Union(Game.BlacksPlayer.Pieces).ToList();

        public Board SetUp()
        {
            for (int i = 1; i <= Size.Width; i++)
            {
                AddWhitePiece(new Pawn(new Position(i, 2)));
                AddBlackPiece(new Pawn(new Position(i, 7)));
            }

            AddWhitePiece(new Rook(new Position("A", 1)));
            AddWhitePiece(new Rook(new Position("H", 1)));
            AddBlackPiece(new Rook(new Position("A", 8)));
            AddBlackPiece(new Rook(new Position("H", 8)));

            AddWhitePiece(new Knight(new Position("B", 1)));
            AddWhitePiece(new Knight(new Position("G", 1)));
            AddBlackPiece(new Knight(new Position("B", 8)));
            AddBlackPiece(new Knight(new Position("G", 8)));

            AddWhitePiece(new Bishop(new Position("C", 1)));
            AddWhitePiece(new Bishop(new Position("F", 1)));
            AddBlackPiece(new Bishop(new Position("C", 8)));
            AddBlackPiece(new Bishop(new Position("F", 8)));

            AddWhitePiece(new Queen(new Position("D", 1)));
            AddBlackPiece(new Queen(new Position("D", 8)));

            AddWhitePiece(new King(new Position("E", 1)));
            AddBlackPiece(new King(new Position("E", 8)));

            return this;
        }

        private void AddWhitePiece(IPiece piece)
        {
            piece.Color = Color.White;
            Game.WhitesPlayer.Pieces.Add(piece);
        }

        private void AddBlackPiece(IPiece piece)
        {
            piece.Color = Color.Black;
            Game.BlacksPlayer.Pieces.Add(piece);
        }
    }
}

File: src/chess-client/src/components/Board.js
import React, { useState } from "react";
import { Row, Col, Container } from "reactstrap";
import Api from "../constants/Api";
import ChessUser from "../common/ChessUser";

const GAME_OVER_TYPE = 2;

const Board = (props) => {
  const [selectedCell, setSelectedCell] = useState("");
  const [isGameOver, setIsGameOver] = useState(false);

  const [positionsInv, positions] = initArrays();

  const { board, gameId } = props;

  const getPieceName = (enumId) => {
    switch (enumId) {
      case 0:
        return "pawn";
      case 1:
        return "rook";
      case 2:
        return "knight";
      case 3:
        return "bishop";
      case 4:
        return "queen";
      case 5:
        return "king";
      default:
        return null;
    }
  };

  const getPiece = (position) => {
    let foundSquare = board.squares.find(
      (square) => square.position.id === position
    );

    if (!foundSquare) return null;

    return foundSquare.piece;
  };

  const renderPiece = (position) => {
    if (!board || !board.squares) return null;

    const foundPiece = getPiece(position);

    if (!foundPiece) return null;

    return (
      <img
        className="piece-img"
        alt={foundPiece.type}
        src={`/chess-pieces/${getPieceName(foundPiece.type)}-${
          foundPiece.color === 0 ? "white" : "black"
        }.svg`}
      ></img>
    );
  };

  const renderCell = (positions, p1) => {
    const getKey = (p2) => cols[p2 - 1].toString() + p1.toString();

    return (
      <>
        {positions.map((p2) => (
          <Col
            key={getKey(p2)}
            className={
              "square " + (selectedCell === getKey(p2) ? "selected" : "")
            }
            onClick={onClickCellHandler.bind(this, getKey(p2))}
          >
            {renderPiece(getKey(p2))}
            <div
              className="position-absolute"
              style={{ left: "1px", bottom: "1px" }}
            >
              {getKey(p2)}
            </div>
          </Col>
        ))}
      </>
    );
  };

  const onClickCellHandler = (positionKey) => {
    if (isGameOver) return;

    if (selectedCell === positionKey) setSelectedCell("");
    else if (selectedCell === "") {
      setSelectedCell(positionKey);
    } else {
      fetch(Api.baseUrl + `/api/games/${gameId}/move`, {
        headers: {
          "Content-Type": "application/json",
          PlayerId: ChessUser.getId(),
        },
        method: "POST",
        body: JSON.stringify({
          origin: selectedCell,
          destination: positionKey,
        }),
      }).then(async (res) => {
        if (res.ok) {
          const data = await res.json();
          const isOver =
            data.turnEvents[data.turnEvents.length - 1].eventType ===
            GAME_OVER_TYPE;
          props.onError("");
          props.onActionPerformed();
          setIsGameOver(isOver);
          setSelectedCell("");
        } else {
          const errMessage = await res.text();
          if (errMessage) {
            props.onError(errMessage);
            setSelectedCell("");
          }
        }
      });
    }
  };

  return (
    <>
      <Container fluid className="board">
        {positionsInv.map((p1, idx1) => (
          <Row key={idx1}>{renderCell(positions, p1)}</Row>
        ))}
      </Container>
    </>
  );
};

const initArrays = () => {
  const positionsInv = [];
  const positions = [];
  for (let i = 8; i >= 1; i--) positionsInv.push(i);
  for (let i = 1; i <= 8; i++) positions.push(i);

  return [positionsInv, positions];
};

const cols = ["A", "B", "C", "D", "E", "F", "G", "H"];

export default Board;
